lib/PrivateAppUserTokenManager.ts

```diff
@@ -0,0 +1,243 @@
+  private tokenMapIntervalId: Map<number, NodeJS.Timeout>;
+  private enabled: boolean;
+
+  constructor(accountId: number) {
+    this.accountId = accountId;
+    this.tokenMap = new Map<number, PrivateAppUserTokenResponse>();
+    this.tokenMapIntervalId = new Map<number, NodeJS.Timeout>();
+    this.enabled = false;
+  }
+
+  async init(): Promise<void> {
+    const scopeGroups = new Set<string>(
+      await scopeGroupsForPersonalAccessKey(this.accountId)
+    );
+    if (
+      !scopeGroups.has(USER_TOKEN_READ) ||
+      !scopeGroups.has(USER_TOKEN_WRITE)
+    ) {
+      throwErrorWithMessage(`${i18nKey}.errors.noScopes`, {
+        accountId: this.accountId,
+      });
+    } else {
+      logger.debug(
+        i18n(`${i18nKey}.enabled`, {
+          accountId: this.accountId,
+        })
+      );
+      this.enabled = true;
+    }
+  }
+
+  cleanup() {
+    this.tokenMapIntervalId.forEach(timeoutId => clearInterval(timeoutId));
+    this.tokenMapIntervalId.clear();
+    this.tokenMap.clear();
+  }
+
+  async getPrivateAppToken(
+    appId: number,
+    scopeGroups: string[] = []
+  ): Promise<string | undefined> {
+    if (!this.enabled) {
+      logger.debug(
+        i18n(`${i18nKey}.disabled`, {
+          accountId: this.accountId,
+          appId: appId,
+        })
+      );
+      return;
+    }
+    try {
+      if (
+        this.tokenMap.has(appId) &&
+        PrivateAppUserTokenManager.doesTokenHaveAllScopes(
+          this.tokenMap.get(appId),
+          scopeGroups
+        )
+      ) {
+        logger.debug(
+          i18n(`${i18nKey}.cached`, {
+            appId: appId,
+          })
+        );
+        return this.tokenMap.get(appId)!.userTokenKey;
+      } else {
+        const token = await this.createOrGetActiveToken(appId, scopeGroups);
+        this.setCacheAndRefresh(appId, token, scopeGroups);
+        return token.userTokenKey;
+      }
+    } catch (err) {
+      let messageDetail = 'Unknown error';
+      if (isAxiosError(err)) {
+        messageDetail = getAxiosErrorWithContext(err as AxiosError).message;
+      } else if (err instanceof Error) {
+        messageDetail = err.message;
+      }
+      throwErrorWithMessage(
+        `${i18nKey}.errors.apiError`,
+        {
+          accountId: this.accountId,
+          appId: appId,
+          messageDetail: messageDetail,
+        },
+        isAxiosError(err) ? undefined : (err as BaseError) // Only pass the error if it's not an AxiosError, api error message is good enough
+      );
+    }
+  }
+
+  isEnabled(): boolean {
+    return this.enabled;
+  }
+
+  private setCacheAndRefresh(
+    appId: number,
+    token: PrivateAppUserTokenResponse,
+    scopeGroups: string[]
+  ) {
+    if (token === undefined || token === null) {
+      throwErrorWithMessage(`${i18nKey}.errors.refreshFailed`, {
+        accountId: this.accountId,
+        appId: appId,
+      });
+    }
+    this.tokenMap.set(appId, token);
```

Comment:
Just riffing, but I wonder if there's a way to avoid needing to have both `tokenMap` and also `tokenMapInterval`.

For example:
```js
this.tokenMap[appId] = {
  value: token,
  refresh: setInterval(() => this.refresh(appId, token), interval)
}
```

Reply:
updated to just one map